Applications/Spire/Source/Ui/SortedTableModel.cpp

```diff
@@ -0,0 +1,283 @@
+      m_mapping[i] += direction;
+    }
+    if(m_reverse_mapping[i] >= row_index) {
+      m_reverse_mapping[i] += direction;
+    }
+  }
+}
+
+std::tuple<int, int> SortedTableModel::find_range(
+    const std::vector<int>::iterator& begin_iterator,
+    const std::vector<int>::iterator& end_iterator, int row) {
+  auto begin = begin_iterator;
+  auto end = end_iterator;
+  for(auto i = m_order.begin(); i != m_order.end(); ++i) {
+    auto new_begin = std::lower_bound(begin, end,
+      m_source->at(row, i->m_index),
+      [&] (int row, const std::any& value) {
+        if(i->m_order == Ordering::ASCENDING) {
+          return m_comparator(m_source->at(row, i->m_index), value);
+        } else {
+          return m_comparator(value, m_source->at(row, i->m_index));
+        }
+      });
+    auto new_end = std::upper_bound(new_begin, end,
+      m_source->at(row, i->m_index),
+      [&] (const std::any& value, int row) {
+        if(i->m_order == Ordering::ASCENDING) {
+          return m_comparator(value, m_source->at(row, i->m_index));
+        } else {
+          return m_comparator(m_source->at(row, i->m_index), value);
+        }
+      });
+    begin = new_begin;
+    end = new_end;
+  }
+  return {static_cast<int>(begin - m_mapping.begin()),
+    static_cast<int>(end - m_mapping.begin())};
+}
+
+void SortedTableModel::on_operation(const Operation& operation) {
+  m_transaction.transact([&] {
+    visit(operation,
+      [&] (const AddOperation& operation) {
+        auto added_index = [&] {
+          if(operation.m_index >= static_cast<int>(m_mapping.size())) {
+            m_mapping.push_back(operation.m_index);
+            m_reverse_mapping.push_back(operation.m_index);
+            return operation.m_index;
+          } else {
+            auto row_index = m_reverse_mapping[operation.m_index];
+            translate(1, operation.m_index);
+            m_mapping.insert(m_mapping.begin() + row_index,
+              operation.m_index);
+            m_reverse_mapping.insert(
+              m_reverse_mapping.begin() + operation.m_index, row_index);
+            return row_index;
+          }
+        }();
+        if(m_order.empty() || m_order.front().m_order == Ordering::NONE) {
+          m_transaction.push(AddOperation{added_index});
+        } else {
+          sort();
+          m_transaction.push(AddOperation{m_reverse_mapping[operation.m_index]});
+        }
+      },
+      [&] (const RemoveOperation& operation) {
+        auto row_index = m_reverse_mapping[operation.m_index];
+        translate(-1, operation.m_index);
+        m_mapping.erase(m_mapping.begin() + row_index);
+        m_reverse_mapping.erase(m_reverse_mapping.begin() + operation.m_index);
+        m_transaction.push(RemoveOperation{row_index});
+      },
+      [&] (const UpdateOperation& operation) {
+        auto row_index = m_reverse_mapping[operation.m_row];
+        m_transaction.push(UpdateOperation{row_index, operation.m_column});
```

Comment:
This violates the requirement that this table is always in sorted order. The update must be performed at the sorted position, not before.

Reply:
The row_index is the index of the sorted table, not the index of the source.